### ISSUE TITLE
LiveReload script is injected inside <header> without warning when <head> is omitted.

### ISSUE THREAD

**Issue by DominoPivot** [gohugoio/hugo#11562]
`livereloadinject.go` naively scans an HTML output as a byte[] and attempts to append a script element:
1.  after the first `>` that follows the first `<head` or `<HEAD`
2.  before `</body>` or `</BODY>`
3.  after the first `>` that follows `<html` or `<HTML` (logs a warning in the browser console)
4.  at the end of the output (logs a warning in the browser console)

<details><summary>Related code</summary>

https://github.com/gohugoio/hugo/blob/1b5f78b6b7335b02b6207a637498c4c8817999d1/transform/livereloadinject/livereloadinject.go#L39-L46 https://github.com/gohugoio/hugo/blob/1b5f78b6b7335b02b6207a637498c4c8817999d1/transform/livereloadinject/livereloadinject.go#L75-L78

</details>

But despite the many fallbacks, **if the output omits the opening `<head>` tag but includes `<head` anywhere else, the script will be injected at an unpredictable location without a warning.** This could be inside a `<header>` element as well as inside a `<style>`, `<script>`, or `<template>` element that happens to contain `<head`.

I discovered this because I typically omit head and body tags, but use a `<header>` for a site's banner. LiveReload gets injected in the header, which has yet to cause issues, but it is still unexpected behavior. I then intentionally made it break a page that contained an inline script with the regex `/#+\s*(?<heading>.+)/g;`.

I believe the LiveReload script should be injected at the start of the head all the time for better performance, but we need a better algorithm to find the implicit start of the head when tags are omitted:
1.  attempt to insert after the `<head>` tag
2.  otherwise attempt to insert after the `<html>` tag
3.  otherwise attempt to insert before the first tag that isn't the `<!doctype>` or a `<!--comment-->`
4.  otherwise return an error

### What version of Hugo are you using (`hugo version`)?

<pre>
$ hugo version
hugo v0.119.0-b84644c008e0dc2c4b67bd69cccf87a41a03937e windows/amd64 BuildDate=2023-09-24T15:20:17Z VendorInfo=gohugoio
</pre>

### Does this issue reproduce with the latest release?
Yes

**Comment by DominoPivot** [comment]
For future reference, even this is a valid HTML document:

```html
<!DocType hTml>
<scrIpt></scriPT>
<TitLe>x</titlE>
```

which produces the same DOM as

```html
<!doctype html>
<html>
  <head>
    <script></script>
    <title>x</title>
  </head>
  <body></body>
</html>
```

**Comment by DominoPivot** [comment]
I've opened two PRs over the last week:
- #11565 followed a similar logic to the existing code but using regular expressions for case insensitivity and a few more use cases. I wasn't happy with this approach as it still failed in some cases and was still not very performant.
- #11578 uses a different approach, consuming the start of the HTML until the start of the head or any unexpected content, and injecting where it stops. This should have better robustness and performance, if there's no mistake in the code.

**Comment by DominoPivot** [comment]
Solved by #11578
